### PR TITLE
Grouping Version 2

### PR DESCRIPTION
**Pull request by Daddelhai**
Arguments can be grouped. Groups have a Header and optional a description. Every argument is in the default Group "arguments", if no other group is assigned.

Fixes Issue #22 
2nd Version.

Example:

```
nitro::broken_options::parser parser("app_name", "about");

parser.group("group1").option("opt", "some opt").short_name("o");
parser.group("group2").toggle("tog", "some toggle").short_name("t");
parser.group("group2").toggle("togg", "some other toggle").short_name("u");
```
OR
```
nitro::broken_options::parser parser("app_name", "about");

auto grp1 = parser.group("group1");
auto grp2 = parser.group("group2");

grp1.option("opt", "some opt").short_name("o");
grp2.toggle("tog", "some toggle").short_name("t");
grp2.toggle("togg", "some other toggle").short_name("u");
```

Output:

```
usage: app_name --opt

about


group1:
  -o [ --opt ] arg                        some opt

group2:
  -t [ --tog ]                            some toggle
  -u [ --togg ]                           some other toggle
```

### CODE REVIEW

**Comment by Daddelhai** [comment]
Unstable at the Moment:
TODO:
- references to map objects
- no sorting
- not enough tests

File: include/nitro/broken_options/option/group.hpp [hunk 1]
```diff
@@ -41,34 +44,189 @@ namespace broken_options
...
+              const std::string& description = std::string(""))
+        : name(name), description(description), all_argument_names_(all_argument_names)
         {
         }
 
-        void add(base& option)
+        group& subgroup(const std::string& name, const std::string& description = std::string(""))
         {
-            options_.emplace(option.name(), std::ref(option));
+            if (std::count(sub_groups_.begin(), sub_groups_.end(), name) == 0)
```
Comment (bmario) [inline include/nitro/broken_options/option/group.hpp:59]: This method should be implemented with only one `find()` and zero `count()` calls.

File: include/nitro/broken_options/option/group.hpp [hunk 2]
```diff
@@ -41,34 +44,189 @@ namespace broken_options
...
+                {
+                    return search->second;
+                }
+            }
+
+            options_.emplace(std::piecewise_construct, std::forward_as_tuple(name),
+                             std::forward_as_tuple(name, description));
+            all_argument_names_.emplace(name);
+
+            return options_.at(name);
```
Comment (bmario) [inline include/nitro/broken_options/option/group.hpp:113]: use the return value of `map::emplace()` instead

File: include/nitro/broken_options/option/group.hpp [hunk 3]
```diff
@@ -41,34 +44,189 @@ namespace broken_options
...
+                {
+                    return search->second;
+                }
+            }
+
+            multi_options_.emplace(std::piecewise_construct, std::forward_as_tuple(name),
+                                   std::forward_as_tuple(name, description));
+            all_argument_names_.emplace(name);
+
+            return multi_options_.at(name);
```
Comment (bmario) [inline include/nitro/broken_options/option/group.hpp:137]: use the return value of `map::emplace()` instead

File: include/nitro/broken_options/option/group.hpp [hunk 4]
```diff
@@ -41,34 +44,189 @@ namespace broken_options
...
+                {
+                    return search->second;
+                }
             }
+
+            toggles_.emplace(std::piecewise_construct, std::forward_as_tuple(name),
+                             std::forward_as_tuple(name, description));
+            all_argument_names_.emplace(name);
+
+            return toggles_.at(name);
```
Comment (bmario) [inline include/nitro/broken_options/option/group.hpp:161]: use the return value of `map::emplace()` instead

File: include/nitro/broken_options/parser.ipp [hunk 5]
```diff
@@ -39,68 +39,78 @@ namespace nitro
...
-        {
-            raise<parser_error>("Trying to redefine toggle as multi_option. Name: ", name);
-        }
+        groups_.emplace_back(group(name, description));
+        return groups_.back();
+    }
+    nitro::broken_options::group& parser::default_group(const std::string& name,
+                                                        const std::string& description)
+    {
+        if (name.length())
```
Comment (bmario) [inline include/nitro/broken_options/parser.ipp:92]: First of all, only Java programmers use `length`, use `size` instead.
And secondly, use `!empty()` .

File: include/nitro/broken_options/option/group.hpp [hunk 6]
```diff
@@ -41,34 +44,190 @@ namespace broken_options
     class group
     {
     public:
-        group(const std::string& name, const std::string& description = std::string(""))
-        : name_(name), description_(description)
+        std::string name;
+        std::string description;
+
+    public:
+        group(std::set<std::string>& all_argument_names, const int& position,
```
Comment (bmario) [inline include/nitro/broken_options/option/group.hpp:51]: Don't pass trivial types by const reference.
```suggestion
        group(std::set<std::string>& all_argument_names, int position,
```

File: include/nitro/broken_options/option/group.hpp [hunk 7]
```diff
@@ -41,34 +44,190 @@ namespace broken_options
...
-        std::string description_;
-        std::map<std::string, std::reference_wrapper<base>> options_;
+        std::set<std::string>& all_argument_names_;
+
+        std::map<std::string, broken_options::option> options_;
+        std::map<std::string, broken_options::multi_option> multi_options_;
+        std::map<std::string, broken_options::toggle> toggles_;
+
+        std::vector<broken_options::group> sub_groups_;
+        int position_;
```
Comment (bmario) [inline include/nitro/broken_options/option/group.hpp:230]: I think `level_` would be a better name for this member.

File: include/nitro/broken_options/option/group.hpp [hunk 8]
```diff
@@ -41,34 +44,190 @@ namespace broken_options
...
+            return toggles_
+                .emplace(std::piecewise_construct, std::forward_as_tuple(name),
+                         std::forward_as_tuple(name, description))
+                .first->second;
+        }
+
+        std::map<std::string, broken_options::option&> get_all_options()
+        {
+            std::map<std::string, broken_options::option&> tmp;
+            for (auto& it : options_)
```
Comment (bmario) [inline include/nitro/broken_options/option/group.hpp:167]: Replace this loop with `tmp.insert(iter, iter)`

File: include/nitro/broken_options/option/group.hpp [hunk 9]
```diff
@@ -41,34 +44,190 @@ namespace broken_options
...
+                auto add = sg.get_all_options();
+                tmp.insert(add.begin(), add.end());
+            }
+            return tmp;
+        }
+
+        std::map<std::string, broken_options::multi_option&> get_all_multi_options()
+        {
+            std::map<std::string, broken_options::multi_option&> tmp;
+            for (auto& it : multi_options_)
```
Comment (bmario) [inline include/nitro/broken_options/option/group.hpp:183]: Replace this loop with `tmp.insert(iter, iter)`

File: include/nitro/broken_options/option/group.hpp [hunk 10]
```diff
@@ -41,34 +44,190 @@ namespace broken_options
...
+            {
+                auto add = sg.get_all_multi_options();
+                tmp.insert(add.begin(), add.end());
+            }
+            return tmp;
+        }
+
+        std::map<std::string, broken_options::toggle&> get_all_toggles()
+        {
+            std::map<std::string, broken_options::toggle&> tmp;
```
Comment (bmario) [inline include/nitro/broken_options/option/group.hpp:198]: Replace this loop with `tmp.insert(iter, iter)`

File: include/nitro/broken_options/option/group.hpp [hunk 11]
```diff
@@ -41,34 +44,190 @@ namespace broken_options
...
+                         std::forward_as_tuple(name, description))
+                .first->second;
+        }
+
+        broken_options::toggle& toggle(const std::string& name,
+                                       const std::string& description = std::string(""))
+        {
+            if (all_argument_names_.find(name) != all_argument_names_.end())
+            {
+                auto search = toggles_.find(name);
```
Comment (bmario) [inline include/nitro/broken_options/option/group.hpp:145]: The result of `find()` shouldn't be called `search`

File: include/nitro/broken_options/option/group.hpp [hunk 12]
```diff
@@ -41,34 +44,190 @@ namespace broken_options
...
+            if (!description.empty())
             {
-                s << std::endl << description_ << std::endl << std::endl;
+                s << std::endl << description << std::endl << std::endl;
             }
 
+            std::map<std::string, const nitro::broken_options::base&> print_options;
             for (auto& iter : options_)
+                print_options.emplace(iter.first, iter.second);
+            for (auto& iter : multi_options_)
```
Comment (bmario) [inline include/nitro/broken_options/option/group.hpp:83]: Use the `insert()` member with two iteratoris instead of the three loops

File: include/nitro/broken_options/parser.hpp [hunk 13]
```diff
@@ -96,11 +109,10 @@ namespace broken_options
         std::string app_name_;
         std::string about_;
 
-        std::map<std::string, broken_options::option> options_;
-        std::map<std::string, broken_options::multi_option> multi_options_;
-        std::map<std::string, broken_options::toggle> toggles_;
+        std::set<std::string> all_argument_names_;
+        std::deque<nitro::broken_options::group> groups_;
```
Comment (bmario) [inline include/nitro/broken_options/parser.hpp:113]: Why is that a deque?

File: include/nitro/broken_options/parser.hpp [hunk 14]
```diff
@@ -38,11 +38,12 @@
 #include <nitro/broken_options/options.hpp>
 
 #include <nitro/except/raise.hpp>
+#include <nitro/lang/unordered.hpp>
```
Comment (bmario) [inline include/nitro/broken_options/parser.hpp:41]: This include seems to be unused here

File: tests/broken_options_test.cpp [hunk 15]
```diff
@@ -243,6 +543,31 @@ about
   -u [ --togg ]                           some other toggle
 )EXPECTED");
     }
+
+    SECTION("Multiple groups with same name should be one group")
```
Comment (bmario) [inline tests/broken_options_test.cpp:547]: The test description doesn't match, what is actually being tested

File: tests/broken_options_test.cpp [hunk 16]
```diff
@@ -243,6 +543,31 @@ about
...
+        std::stringstream s;
+
+        auto grp1 = parser.group("group1");
+        auto grp2 = parser.group("group2");
+
+        grp1.toggle("tog", "should work");
+        REQUIRE_THROWS_AS(grp2.toggle("tog", "should fail"), nitro::broken_options::parser_error);
+    }
+
+    SECTION("Groups work 2")
```
Comment (bmario) [inline tests/broken_options_test.cpp:560]: Improve that test description

File: tests/broken_options_test.cpp [hunk 17]
```diff
@@ -208,13 +208,313 @@ TEST_CASE("Usage descriptions work")
...
+            }
+        }
+
+        REQUIRE(actual == expected);
+    }
+}
+
+TEST_CASE("Groups work")
+{
+    SECTION("Groups work")
```
Comment (bmario) [inline tests/broken_options_test.cpp:268]: Improve test description

File: tests/broken_options_test.cpp [hunk 18]
```diff
@@ -208,13 +208,313 @@ TEST_CASE("Usage descriptions work")
...
+                ++tmp;
+                if (tmp > 100)
+                    break;
+            }
+        }
+
+        REQUIRE(actual == expected);
+    }
+
+    SECTION("Groups work 2")
```
Comment (bmario) [inline tests/broken_options_test.cpp:368]: Improve test description

File: tests/broken_options_test.cpp [hunk 19]
```diff
@@ -208,13 +208,313 @@ TEST_CASE("Usage descriptions work")
...
+                ++tmp;
+                if (tmp > 100)
+                    break;
+            }
+        }
+
+        REQUIRE(actual == expected);
+    }
+
+    SECTION("Change Default Group")
```
Comment (bmario) [inline tests/broken_options_test.cpp:468]: Improve test description

File: tests/broken_options_test.cpp [hunk 20]
```diff
@@ -243,6 +543,31 @@ about
...
+    SECTION("Groups work 2")
+    {
+        nitro::broken_options::parser parser("app_name", "about");
+
+        std::stringstream s;
+
+        auto grp1 = parser.group("group1");
+        auto grp2 = parser.group("group1");
+
+        REQUIRE(grp1 == grp2);
```
Comment (bmario) [inline tests/broken_options_test.cpp:569]: This condition looks okay, but what about (pseudo-code):

```
auto opt1 = grp1.option("opt");
auto opt2 = grp2.option("opt");

REQUIRE(opt1 == opt2);
```

**Review by bmario — CHANGES_REQUESTED** [review]

File: tests/broken_options_test.cpp (hunk 20 above)
Comment (Daddelhai) [reply tests/broken_options_test.cpp:569]: `auto opt2 = grp2.option("opt");` raises an exception because 'opt' is already defined in grp1

File: include/nitro/broken_options/option/group.hpp (hunk 8 above)
Comment (Daddelhai) [reply include/nitro/broken_options/option/group.hpp:166]: Impossible:
Cast from broken_options::option to broken_options::option&

File: include/nitro/broken_options/option/group.hpp (hunk 9 above)
Comment (Daddelhai) [reply include/nitro/broken_options/option/group.hpp:180]: same here

File: include/nitro/broken_options/option/group.hpp (hunk 10 above)
Comment (Daddelhai) [reply include/nitro/broken_options/option/group.hpp:193]: same here

File: include/nitro/broken_options/option/group.hpp (hunk 12 above)
Comment (Daddelhai) [reply include/nitro/broken_options/option/group.hpp]: Impossible due cast

File: include/nitro/broken_options/parser.hpp (hunk 13 above)
Comment (Daddelhai) [reply include/nitro/broken_options/parser.hpp:112]: I thought this is the best option... :/

File: tests/broken_options_test.cpp (hunk 20 above)
Comment (Daddelhai) [reply tests/broken_options_test.cpp]: This works well, if you use auto& grpX ...
I also added a test for this case.

File: include/nitro/broken_options/parser.hpp (hunk 13 above)
Comment (bmario) [reply include/nitro/broken_options/parser.hpp:112]: Let me rephrase that: Why are you using a `deque` instead of a `vector`?

File: tests/broken_options_test.cpp [hunk 21]
```diff
@@ -208,13 +208,313 @@ TEST_CASE("Usage descriptions work")
...
+            }
+        }
+
+        REQUIRE(actual == expected);
+    }
+}
+
+TEST_CASE("Groups")
+{
+    SECTION("Groups output is correct")
```
Comment (bmario) [inline tests/broken_options_test.cpp:268]: Which output is correct? Please improve further

File: tests/broken_options_test.cpp [hunk 22]
```diff
@@ -208,13 +208,313 @@ TEST_CASE("Usage descriptions work")
...
+                ++tmp;
+                if (tmp > 100)
+                    break;
+            }
+        }
+
+        REQUIRE(actual == expected);
+    }
+
+    SECTION("Groups output is correct #2")
```
Comment (bmario) [inline tests/broken_options_test.cpp:368]: Which output is correct and why does it need a second test case? Please improve the description further.

File: tests/broken_options_test.cpp [hunk 23]
```diff
@@ -208,13 +208,313 @@ TEST_CASE("Usage descriptions work")
...
+    {
+        nitro::broken_options::parser parser("app_name", "about");
+
+        std::stringstream s;
+
+        parser.default_group("test arguments");
+        parser.accept_positionals(3);
+        parser.positional_name("command line");
+
+        parser.toggle("tog", "some toggle").short_name("t");
```
Comment (bmario) [inline tests/broken_options_test.cpp:478]: So the test shall only check the name of the default group. Reduced the test to the bare minimum required to test the name

**Review by bmario — CHANGES_REQUESTED** [review]

File: include/nitro/broken_options/parser.hpp (hunk 13 above)
Comment (Daddelhai) [reply include/nitro/broken_options/parser.hpp:112]: I can't use a vector, because the objects in a vector can be relocated and there are references in to the groups_ list :/

File: tests/broken_options_test.cpp [hunk 24]
```diff
@@ -241,8 +243,166 @@ about
...
+                                          very very very very very very very
+                                          very very very very very very very
+                                          very very very very very very very
+                                          very very very very very long
+                                          description
+)EXPECTED";
+
+        CHECK(actual.size() == expected.size());
+
+        std::size_t tmp = 0;
```
Comment (bmario) [inline tests/broken_options_test.cpp:356]: Please remove this debugging code, the test should only compare

File: tests/broken_options_test.cpp [hunk 25]
```diff
@@ -241,8 +243,166 @@ about
                                           description
   -t [ --tog ]                            some toggle
   -u [ --togg ]                           some other toggle
+)EXPECTED";
+
+        CHECK(actual.size() == expected.size());
+
+        std::size_t tmp = 0;
```
Comment (bmario) [inline tests/broken_options_test.cpp:250]: Please remove the debugging code. Ideally, this test shouldn't change at all. (Except for the newline)

**Review by bmario — CHANGES_REQUESTED** [review]
Please prioritize this PR, so we can merge it soon.

**Review by bmario — APPROVED** [review]
